src/darwin/Framework/CHIPTests/MTROTAProviderTests.m

```diff
@@ -888,7 +925,482 @@ - (void)test005_DoBDXTransferAllowUpdateRequest
+        NSString * otaImageFilePath = (isDeviceID1) ? otaRawImagePath1 : otaRawImagePath2;
+        NSString * otaDownloadedFilePath = (isDeviceID1) ? kOtaDownloadedFilePath1 : kOtaDownloadedFilePath2;
+
+        XCTAssertEqual(controller, sController);
+        XCTAssertEqualObjects(params.updateToken, updateToken);
+        XCTAssertEqualObjects(params.newVersion, kSoftwareVersion);
+
+        XCTAssertTrue([[NSFileManager defaultManager] contentsEqualAtPath:otaImageFilePath andPath:otaDownloadedFilePath]);
+
+        [sOTAProviderDelegate respondToApplyUpdateRequestWithAction:MTROTASoftwareUpdateProviderOTAApplyUpdateActionProceed
+                                                         completion:completion];
+
+        if (isDeviceID1) {
+            [applyUpdateRequestExpectation1 fulfill];
+        } else {
+            [applyUpdateRequestExpectation2 fulfill];
+        }
+    };
+
+    sOTAProviderDelegate.notifyUpdateAppliedHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+        MTROTASoftwareUpdateProviderClusterNotifyUpdateAppliedParams * params, MTRStatusCompletion completion) {
+        XCTAssertTrue([nodeID isEqualToNumber:@(kDeviceId1)] || [nodeID isEqualToNumber:@(kDeviceId2)]);
+
+        bool isDeviceID1 = [nodeID isEqualToNumber:@(kDeviceId1)];
+        NSData * updateToken = (isDeviceID1) ? updateToken1 : updateToken2;
+
+        // Device1 is updated to version 10 and device2 to version 5.
+        NSNumber * kSoftwareVersion = (isDeviceID1) ? kUpdatedSoftwareVersion_10 : kUpdatedSoftwareVersion_5;
+
+        XCTAssertEqual(controller, sController);
+        XCTAssertEqualObjects(params.updateToken, updateToken);
+        XCTAssertEqualObjects(params.softwareVersion, kSoftwareVersion);
+
+        [sOTAProviderDelegate respondSuccess:completion];
+
+        if (isDeviceID1) {
+            hasDevice1NotifiedUpdateApplied = true;
+            [notifyUpdateAppliedExpectation1 fulfill];
+        } else {
+            hasDevice2NotifiedUpdateApplied = true;
+            [notifyUpdateAppliedExpectation2 fulfill];
+        }
+
+        // If both device1 and device have notified that the update has been applied, reset the
+        // applyUpdateRequestHandler and notifyUpdateAppliedHandler to nil
+        if (hasDevice1NotifiedUpdateApplied && hasDevice2NotifiedUpdateApplied) {
+            sOTAProviderDelegate.applyUpdateRequestHandler = nil;
+            sOTAProviderDelegate.notifyUpdateAppliedHandler = nil;
+        }
+    };
+
+    // Advertise ourselves as an OTA provider.
+    XCTestExpectation * announceResponseExpectation1 = [self announceProviderToDevice:device1];
+
+    // Make sure we get our callbacks in order for both device1 and device2 since we do not
+    // send image available to device2 until BDX for device1 has ended so queryExpectation3 can follow
+    // bdxEndExpectation1. Give it a bit more time, because we want to allow time for the BDX download.
```

Comment:
```suggestion
    // Make sure we get our callbacks in order for both device1 and device2.  Since we do not
    // send image available to device2 until BDX for device1 has ended, queryExpectation3 must follow
    // bdxEndExpectation1.
    // 
    // Give it a bit more time, because we want to allow time for the BDX downloads.
```